### DISCUSSION TITLE
Is learning Carbon is hard as compared to C++?

### DISCUSSION THREAD

**Discussion by prathmesh-b** [carbon-language/carbon-lang#1822] [Explorer]
Carbon is just launched by google. So I wonder is it going to be hard to learn as compared to C++

**Comment by KateGregory** [comment]
Right now, it is impossible to learn Carbon. It isn't done. There is no compiler. We are looking for people to join the project and finish the language design and implementation, plus documentation and more.

Long term, a goal is to be easier to learn and easier to teach than C++. Some things in C++ are complicated for historical reasons. Things were done a certain way 30 years ago, then a keyword was added, and these days best practices are to always use that keyword, but it can't be the default because that would break old code. So we have to teach people to remember to use the keyword all the time. In Carbon, we have the opportunity to make the simpler and more modern choice, and not have to teach people about the "sharp edges" where if you don't know, you might do something you shouldn't.

**Reply by OlaFosheimGrostad** [reply]
> Some things in C++ are complicated for historical reasons. Things were done a certain way 30 years ago, then a keyword was added, and these days best practices are to always use that keyword, but it can't be the default because that would break old code.

I think it is possible to write a new syntax for C++ that compiles to the same abstract syntax tree, but macros make upgrading the syntax very difficult.

Do you know if Carbon will use any kind of rewrite-system that could give the same limitations? (AST macros have the same issue, any change is a potential breaking change.)

**Reply by YagaoDirac** [reply]
> Long term, a goal is to be easier to learn and easier to teach than C++. 

It's about the production over cost. It's not about purely reducing the cost. Make sure Carbon does as much as needed. If it really cost a lot to learn, it should still be at least ok. Don't simply the language too much. 
Btw, please consider add auto completion to language feature. Don't leave it for IDEs to implement.
